pkg/monitor/intervalcreation/operator.go

```diff
@@ -0,0 +1,87 @@
+package intervalcreation
+
+import (
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+)
+
+func IntervalsFromEvents_OperatorAvailable(events []*monitorapi.Event, beginning, end time.Time) monitorapi.EventIntervals {
+	return intervalsFromEvents_OperatorStatus(events, beginning, end, configv1.OperatorAvailable, configv1.ConditionTrue, monitorapi.Error)
+}
+
+func IntervalsFromEvents_OperatorProgressing(events []*monitorapi.Event, beginning, end time.Time) monitorapi.EventIntervals {
+	return intervalsFromEvents_OperatorStatus(events, beginning, end, configv1.OperatorProgressing, configv1.ConditionFalse, monitorapi.Warning)
+}
+
+func IntervalsFromEvents_OperatorDegraded(events []*monitorapi.Event, beginning, end time.Time) monitorapi.EventIntervals {
+	return intervalsFromEvents_OperatorStatus(events, beginning, end, configv1.OperatorDegraded, configv1.ConditionFalse, monitorapi.Error)
+}
+
+func intervalsFromEvents_OperatorStatus(events []*monitorapi.Event, beginning, end time.Time, conditionType configv1.ClusterStatusConditionType, conditionGoodState configv1.ConditionStatus, level monitorapi.EventLevel) monitorapi.EventIntervals {
+	ret := monitorapi.EventIntervals{}
+	operatorToInterestingBadCondition := map[string]*configv1.ClusterOperatorStatusCondition{}
+
+	for _, event := range events {
+		operatorName, ok := monitorapi.OperatorFromLocator(event.Locator)
+		if !ok {
+			continue
+		}
+		currentCondition := monitorapi.GetOperatorConditionStatus(event.Message)
+		if currentCondition == nil {
+			continue
+		}
+		if currentCondition.Type != conditionType {
+			continue
+		}
+
+		lastCondition := operatorToInterestingBadCondition[operatorName]
+		if lastCondition != nil && lastCondition.Status == currentCondition.Status {
+			// if the status didn't actually change (imagine degraded just changing reasons)
+			// don't count as the interval
+			continue
+		}
+		if currentCondition.Status != conditionGoodState {
+			// don't overwrite a previous condition in a bad state
+			if lastCondition == nil {
+				// force teh last transition time, sinc we think we just transitioned at this instant
+				currentCondition.LastTransitionTime.Time = event.At
+				operatorToInterestingBadCondition[operatorName] = currentCondition
+			}
+			continue
+		}
+
+		// at this point we have transitioned to a good state.  Remove the previous "bad" state
+		delete(operatorToInterestingBadCondition, operatorName)
+
+		from := beginning
+		if lastCondition != nil {
+			from = lastCondition.LastTransitionTime.Time
+		}
+		ret = append(ret, &monitorapi.EventInterval{
+			Condition: &monitorapi.Condition{
+				Level:   level,
+				Locator: event.Locator,
+				Message: fmt.Sprintf("condition/%s status/%s reason/%s", conditionType, lastCondition.Status, lastCondition.Message),
```

Comment:
/hold

[metal-ipi][1]:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1dc4691]
goroutine 1 [running]:
github.com/openshift/origin/pkg/monitor/intervalcreation.intervalsFromEvents_OperatorStatus(0xc007280000, 0x57af, 0x6000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x5b84c56, ...)
	github.com/openshift/origin/pkg/monitor/intervalcreation/operator.go:67 +0x431
github.com/openshift/origin/pkg/monitor/intervalcreation.IntervalsFromEvents_OperatorProgressing(0xc007280000, 0x57af, 0x6000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x92b6fc8, ...)
	github.com/openshift/origin/pkg/monitor/intervalcreation/operator.go:16 +0xdb
github.com/openshift/origin/pkg/monitor.(*Monitor).EventIntervals(0xc001264a10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x1, 0x0)
	github.com/openshift/origin/pkg/monitor/monitor.go:141 +0x2f7
...
```

You have a `lastCondition != nil` comparison above, so you need some sort of guard against `lastCondition == nil` here.

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/26034/pull-ci-openshift-origin-master-e2e-metal-ipi-ovn-ipv6/1377620127620534272#1:build-log.txt%3A14596

Reply:
> You have a `lastCondition != nil` comparison above, so you need some sort of guard against `lastCondition == nil` here.

thanks.  Fixed.  Reapplied lgtm and removed hold